lib/services/protocol_service.ts

```diff
@@ -572,11 +572,30 @@ export class SNProtocolService extends PureService implements EncryptionDelegate
         continue;
       }
       try {
+        let keyForPayload = key;
+        const version = encryptedPayload!.version;
+        const payloadContentType = encryptedPayload!.content_type;
+        const keyContentType = key?.content_type;
+        /** We check if the key can decrypt the encrypted payload. */
+        if (
+          version === ProtocolVersion.V004 &&
+          keyContentType === ContentType.RootKey &&
+          payloadContentType !== ContentType.ItemsKey
+        ) {
+          keyForPayload = undefined;
+        }
         const decryptedPayload = await this.payloadByDecryptingPayload(
           encryptedPayload,
-          key
+          keyForPayload
         );
-        decryptedPayloads.push(decryptedPayload);
+        /**
+         * If waitingForKey, we must import the payload in order to decrypt it later.
+         */
+        if (decryptedPayload.waitingForKey) {
+          await this.modelManager!.importPayloads([decryptedPayload]);
```

Comment:
This is no good. This function should not have any side effects. It should be functional. Pretend it is a utility method. It should not trigger state updates.

Reply:
(This is regarding `await this.modelManager!.importPayloads([decryptedPayload]);`. Should not be doing this here.)

Reply:
What about inside `importData`: https://github.com/standardnotes/snjs/blob/a27979314775f7aa34d4817bee740ab3fa3948c5/lib/application.ts#L759

We can import items there before importing valid items.

Reply:
That would seem to make sense, but I'd have to see the solution in its final form to see how it plays out.

Reply:
Done 👍